doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md

```diff
@@ -0,0 +1,530 @@
+# SONiC GNMI Server Interface Design
+
+# High Level Design Document
+
+#### Rev 0.1
+
+# Table of Contents
+- [Table of Contents](#table-of-contents)
+- [List of Tables](#list-of-tables)
+- [Revision](#revision)
+- [About this Document](#about-this-document)
+- [Scope](#scope)
+- [Definition/Abbreviation](#definitionabbreviation)
+- [1 Feature Overview](#1-feature-overview)
+  * [1.1 Requirements](#11-requirements)
+  * [1.2 Design Overview](#12-design-overview)
+- [2 Error Handling](#2-error-handling)
+- [3 Serviceability and Debug](#3-serviceability-and-debug)
+- [4 Unit Tests](#4-unit-tests)
+  * [4.1 Unit Tests for Capabilities RPC](#41-unit-tests-for-capabilities-rpc)
+  * [4.2 Unit Tests for Get RPC](#42-unit-tests-for-get-rpc)
+  * [4.3 Unit Tests for Set RPC](#43-unit-tests-for-set-rpc)
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+[Table 2: Mapping Operations](#table-2-mapping-operations)
+
+[Table 3: Translate Example](#table-3-translate-example)
+
+# Revision
+
+| Rev | Date        | Author             | Change Description  |
+|:---:|:-----------:|:------------------:|---------------------|
+| 0.1  | 05/05/2022 | Gang Lv    | Initial version     |
+
+# About this Document
+This document provides a detailed description on the strategy to implement the SONiC GNMI Server Interface.
+
+# Scope
+This document describes the high level design of SONiC GNMI Server Interface. 
+
+This document provides minor implementation details about the proposed solutions.
+
+# Definition/Abbreviation
+
+### Table 1: Abbreviations
+| **Term** | **Meaning**                |
+| -------- | -------------------------- |
+| gRPC     | A modern open-source high performance Remote Procedure Call (RPC) framework that can run in any environment |
+| gNMI     | gRPC Network Management Interface, used to retrieve or manipulate the state of a device via telemetry or configuration data |
+| gNOI     | gRPC Network Operations Interface |
+| JSON     | JavaScript Object Notation |
+| Yang     | Yet Another Next Generation, it is a modular language representing data structures in an XML tree format |
+| Generic Config Updater | SONiC configuration generic update and rollback feature |
+| ConfigDB | Configuration Database |
+| ApplDB | Application Database |
+
+# 1 Feature Overview
+
+SONiC should provide a gNMI/gNOI server interface where the client can communicate with SONiC (server) through the interface using their respective gRPC libraries based on ConfigDB/ApplDB/StateDB/CountersDB schema or SONiC Yang schema.
```

Comment:
This was discussed in the meeting. There are some usecases such as route add/deletes (for e.g vnet routes) which shall be programmed by an external controller. These routes are not intended to be persistent across reboots. It is essentially extending the existing RestAPI model to gNMI.

Reply:
@prsunny We need to selectively allow (e.g VNET routes, DASH objects in APPL_DB) to control via REST/gNMI server, exposing DB objects via north bound interfaces would be a big security concern. I like to see some design around this in this HLD or in a separate HLD.

Reply:
@prsunny I don't think we can make infrastructure changes to allow specific need for a select-isolated deployment. Tomorrow someone might comeup with a need to directly write to ASIC_DB for effecting shutdown of all port-channels quickly. I don't think these things should be allowed. I don't agree that 'some external entity' wants to write to internal database tables of SONIC' is a requirements that needs to be fulfilled.

Reply:
Agree with Venkat that it shall be only selectively allowed. @bandaru-viswanath, i tend to disagree that this is infrastructure change. For your comment, if there is an external orchagent, may be it is a valid scenario of writing directly to ASIC_DB. But currently that's not the case. As mentioned, this is not a new design as APP_DB means application DB and depending on where the application sits (like BGP etc), it can act as the producer. Why should we restrict that application must be localized. Also this is not a persitent config. So config_db is out of scope.

Reply:
Please suggest how it can be selectively enabled, without hardcoding specific details in the container. The only place this can be done is in the GNMI container, and if we do so, we will be hardcoding some specific usecase/deployment requirements into UI.

As suggested above, this can be done using Yang. 

Yes, APPL_DB is the right place. Im just saying config_db is out of scope for programming something that is not intended to be persistent. I'm not sure on how positioning of APPL database in architecture warrants that only config_db can be modified for write operations.  My point is APPL_DB is for application data (irrespective of whether it is local or remote). Agree with you on ASIC_DB.

Reply:
Just a simple clarification: there is no requirement to write directly to STATE_DB, in this HLD, the gNMI server will read from StateDB, like a telemetry server.